consensus/src/sync/live/block_queue/mod.rs

```diff
@@ -304,12 +301,81 @@ impl<N: Network> BlockQueue<N> {
         );
     }
 
+    /// Attempts to insert a block into the buffer.
+    /// This may fail due to the peer who propagated the block as given in the optional `pubsub_id` parameter
+    /// already having too many blocks buffered. Blocks without a pubsub_id will always be buffered.
+    ///
+    /// At most two block per peer per block height will be buffered leaving enough room for a bock and the
+    /// skip block making it obsolete. For benign peers that is sufficient, as they will never propagate
+    /// more than that.
+    ///
+    /// ## Returns
+    /// * `true` if the block was added to the buffer
+    /// * `false` otherwise
     fn insert_block_into_buffer(&mut self, block: Block, pubsub_id: Option<N::PubsubId>) -> bool {
-        self.buffer
-            .entry(block.block_number())
-            .or_default()
-            .insert(block.hash(), (block, pubsub_id))
-            .is_some()
+        let peer_id = match pubsub_id {
+            // Blocks without a pubsub_id are requested and will always be buffered.
+            None => {
+                return self
+                    .buffer
+                    .entry(block.block_number())
+                    .or_default()
+                    .insert(block.hash(), (block, pubsub_id))
+                    .is_none()
+            }
+            // For blocks with a pubsub id the propagation source is the peer which can
+            // only have 2 items buffered.
+            Some(ref pubsub_id) => pubsub_id.propagation_source(),
+        };
+
+        // Get the entry for the block number if it exists.
+        // Otherwise add the entry and add the block to it and return.
+        let mut entry = match self.buffer.entry(block.block_number()) {
+            BTreeMapEntry::Occupied(occupied_entry) => occupied_entry,
```

Comment:
```suggestion
            BTreeMapEntry::Occupied(occupied_entry) => occupied_entry.into_mut(),
```

Using `into_mut()` instead of `get_mut()` below should get rid of the borrow checker problems.